wal/src/message_queue_impl/region.rs

```diff
@@ -200,13 +252,260 @@ impl<M: MessageQueue> Region<M> {
+
+        loop {
+            let (latest_message_and_offset, returned_high_watermark) = iter
+                .next_message()
+                .await
+                .map_err(|e| Box::new(e) as _)
+                .context(OpenWithCause {
+                    namespace,
+                    region_id,
+                    msg: "failed while recover from log",
+                })?;
+
+            // TODO: maybe should assert it?
+            ensure!(returned_high_watermark == high_watermark, OpenNoCause { namespace , region_id, msg: format!(
+                "failed while recover from log, high watermark shouldn't changed while opening region, 
+                origin high watermark:{}, returned high watermark:{}", high_watermark, returned_high_watermark)
+            });
+
+            // Decode and apply it to builder.
+            let raw_key = latest_message_and_offset
+                .message
+                .key
+                .with_context(|| OpenNoCause {
+                    namespace,
+                    region_id,
+                    msg: "failed while recover from log, key in message shouldn't be None",
+                })?;
+
+            let raw_value =
+                latest_message_and_offset
+                    .message
+                    .value
+                    .with_context(|| OpenNoCause {
+                        namespace,
+                        region_id,
+                        msg: "failed while recover from log, value in message shouldn't be None",
+                    })?;
+
+            let key = log_encoding
+                .decode_key(raw_key.as_slice())
+                .map_err(|e| Box::new(e) as _)
+                .context(OpenWithCause {
+                    namespace,
+                    region_id,
+                    msg: "failed while recover from log",
+                })?;
+
+            ensure!(key.region_id == region_id, OpenNoCause { namespace , region_id, msg: format!(
+                "failed while recover from log, region id in key should be equal to the one of current region,
+                but now are {} and {}", key.region_id, region_id)
+            });
+
+            // TODO: maybe this clone should be avoided?
+            let region_meta_delta = RegionMetaDelta::new(
+                key.table_id,
+                key.sequence_num,
+                latest_message_and_offset.offset,
+            );
+
+            builder
+                .apply_region_meta_delta(region_meta_delta.clone())
+                .map_err(|e| Box::new(e) as _)
+                .context(OpenWithCause {
+                    namespace,
+                    region_id,
+                    msg: format!(
+                        "failed while recover from log, region meta delta:{:?}",
+                        region_meta_delta
+                    ),
+                });
+
+            // Has polled last log in topic, break.
+            if latest_message_and_offset.offset + 1 == high_watermark {
```

Comment:
Any possibilities that the `latest_message_and_offset.offset + 1` always larger than `high_watermark`? Shall we do some kind of check for this case?

Reply:
There has been check above to ensure `high watermark` equal to `returned high watermark`, and `offset + 1` will always be less than or equal to `returned high watermark`.  I think checking again may be unnecessary?